source/4_tools/regina/src/regina.cpp

```diff
@@ -1,54 +1,865 @@
+        g_regina_ctx.proj = glm::transpose(g_regina_ctx.proj); // DirectX backend ( so we have to transpose, expects row major matrices )
+
+        return true;
+    }
+    //-------------------------------------------------------------------------
+    void update()
+    {
+        update_view();
+
+        // Has the GPU finished processing the commands of the current frame resource?
+        // If not, wait until the GPU has completed commands up to this fence point.
+        // 
+        // This call is not queued but directly executed!
+        renderer::wait_for_active_frame();
+    }
+    //-------------------------------------------------------------------------
+    void draw()
+    {
+        Viewport vp = { 0.0f, 0.0f, (f32)globals::window_info().width, (f32)globals::window_info().height, 0.0f, 1.0f};
+        ScissorRect sr = { vp.top_left_x, vp.top_left_y, vp.width, vp.height };
+
+        renderer::set_pipeline_state_object(g_regina_ctx.pso);
+        renderer::set_raster_state(g_regina_ctx.solid_raster_state);
+
+        renderer::new_frame();
+
+        update_object_constant_buffers();
+        update_pass_constant_buffers();
+
+        renderer::set_render_targets(globals::default_targets_info().back_buffer_color , globals::default_targets_info().depth_buffer);
+        renderer::set_viewport(vp);
+        renderer::set_scissor_rect(sr);
+
+        renderer::begin_draw();
+
+        renderer::clear(g_regina_ctx.clear_state);
+
+        renderer::set_shader(g_regina_ctx.shader_program);
+
+        renderer::ResourceSlot curr_pass_cb = get_active_pass_constant_buffer_for_frame(renderer::active_frame()->slot_id());
+        renderer::set_constant_buffer_view(curr_pass_cb, 1);
+
+        for (auto& ri : (*g_regina_ctx.scene))
+        {
+            renderer::set_vertex_buffer(ri.geometry->vertex_buffer, 0, ri.geometry->vertex_byte_stride, 0);
+            renderer::set_index_buffer(ri.geometry->index_buffer, renderer::IndexBufferFormat::R16_UINT, 0);
+            renderer::set_primitive_topology(renderer::PrimitiveTopology::TRIANGLELIST);
+
+            renderer::ResourceSlot curr_object_cb = get_active_object_constant_buffer_for_frame(renderer::active_frame()->slot_id(), ri.constant_buffer_index);
+            renderer::set_constant_buffer_view(curr_object_cb, 0);
+
+            renderer::renderer_draw_indexed_instanced(1, 0, ri.index_count, ri.start_index_location, ri.base_vertex_location);
+        }
+
+        renderer::end_draw();
+
+        renderer::present();
+        renderer::end_frame();
+    }
+    //-------------------------------------------------------------------------
+    void shutdown()
+    {
+        REX_LOG(LogRegina, "shutting down Regina");
+
+        g_regina_ctx.clear_state.release();
```

Comment:
you shouldn't have to release these manually.
also I think they should be managed by the scene, not by regina itself

Reply:
Once I convert them into pointers instead of shared copies we do not no, but now this hold a reference to back buffer render targets. if we don't release these manually they will get cleared after main, when all allocated memory has already been destroyed, this will result in a crash. I have made some changes already to the ResourceSlots but I'm going to convert them into pointers into another PR as this one is already big enough.

Reply:
linked to https://github.com/New-Creta/Rex/issues/268